Improve performance and correctness of Tiling Patterns

Rob--W opened the pull request:
This PR significantly improves the performance of PDFs with tiling patterns (which also fixes a bunch of blurry images), and fixes a correctness issue caused by a missing transform.

Fixes #6496
Fixes #5698
Fixes #1434
Fixes #2825 (the rendering issues have been fixed, but there are still spikes in memory usage (which is very problematic on low-memory devices), this is probably caused by https://crbug.com/448150, and possibly the same issue as https://github.com/bp74/StageXL/issues/214?. I think that we can work around this memory spike by introducing an artificial delay in our task scheduler to give the browser a chance to GC the temporary CanvasGradient patterns - this could be added in a separate PR).

Fixes #6737
Fixes #7731
Fixes #8264
Fixes #8579
Fixes #9922
Fixes #9927
Fixes #9928
Fixes #7946
Fixes #8083
Fixes #8373
Fixes #9770
Fixes #9665
Fixes #9763
Fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1467424

Rob--W committed:
Refactor TilingPattern

- Deduplicate size/scale calculation, by introducing `getSizeAndScale`.
- Eliminate unnecessary calculations / variables.

Rob--W committed:
Avoid extreme sizing / scaling in tiling pattern

The new test file (tiling-pattern-large-steps.pdf) was manually created,
to have the following characteristics:
- Large xstep and ystep (90000)
- Page width is 4000 (which is larger than MAX_PATTERN_SIZE)
- Visually, the page consists of a red rectangle with a black border,
  surrounded by a 50 unit white padding.
- Before patch: blurry; After patch: sharp

Fixes #6496
Fixes #5698
Fixes #1434
Fixes #2825

Rob--W committed:
TilingPattern: Reverse transform after painting

This transform resulted in an incorrectly positioned object when the
bounding box's upper-left corner did not start at (0,0), because
the translation was not reverted. This patch adds the missing transform.

The test file (tiling-pattern-box.pdf) is based on the PDF from #2825.
All but the first cube (including the PDF data) have been removed.
To trigger the bug that is fixed by this commit, I changed the BBox of
the first pattern from "[ 0 0 596 842]" to "[90 0 596 842]". Without
this patch, the dashed vertical line that intersects the corners at A
and E would disappear.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Amazing, thank you for looking into this! I can confirm that all four issues are fixed with this PR. The PDF file in the last issue is rendering much faster (almost instantly for me) now and I don't really experience the mentioned memory spike, so I think we're fine there too.

/botio test

Rob--W commented:
@timvandermeij I've found the memory spike in Chrome (on Linux), using the following steps:

1. Download the PDF file from https://github.com/mozilla/pdf.js/issues/2825#issuecomment-24234561 (to make sure that rendering happens at once, without waiting for network)
2. Open the PDF.js PDF Viewer.
3. Open the task manager of Chrome (Shift + Esc) and keep an eye on the tab with PDF.js
4. In the Open PDF dialog, select the PDF file that you downloaded from step 1.
5. While the PDF is being rendered, memory usage shoots up to over 1 GB. Shortly after completion of rendering, the memory usage drops again.

Rob--W commented:
The test previously failed with observable differences because I accidentally used `Math.max` instead of `Math.min`

I have updated the PR with the fix. When I ran the tests locally, 3 tests failed (chrome-High-Pressure-Measurement-WP-001287-page3,  chrome-colorkeymask-page1, chrome-issue2177-eq-page1). These failures are caused by slight differences in rounding errors (previously we manually calculated the scale, now most of the scale calculations is delegated to the canvas). I don't see any visual problems with the rendering, so you have to regenerate the refs for these three test cases after merging this PR.

Snuffleupagus commented:
/botio test

yurydelendik commented on src/display/pattern_helper.js:
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
-      var botRight = [x0 + xstep, y0 + ystep];
+      // It is a waste of memory to use the size as specified by xstep or ystep
+      // if the bounding box is smaller.
+      xstep = Math.min(xstep, x1);
```
x0 and x1 are left and right coordinates of the box. shall it be `x1 - x0` here?

Rob--W replied:
I don't think so (correct me if I'm wrong):
The drawing is clipped below (by `clipBbox`) by the square (x0,y0) - (x1,y1). So the canvas itself must be at least as large as (x1,y1).

Using the difference `x1 - x0` would be possible if we can translate the viewport and clip at the rectangle (0,0) - (x1-x0, y1-y0). I haven't tried this yet, do you think that this is a valid operation?

yurydelendik replied:
yeah, that's what translation -x0, -y0 below shall do... let me check why fad83a7 fixes the issue

yurydelendik replied:
lots of mistakes was in the initial code, let's see if we can fix them:
- we shall rely on pattern cell width/height not on its clip boundary coordinate or xstep -- the mistake was we used xstep instead of pattern width, I don't think xstep shall participate in the calculations
- once we translated by -x0,-y0, we shall do the reverse before painting (at the scaleToContext spot)

yurydelendik replied:
See also http://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/PDF32000_2008.pdf#page=183&zoom=180,23,420 

Also notice XStep/YStep comments below:
- "XStep  and  YStep  may  differ  from  the  dimensions  of  the  pattern  cell  implied  by  the  BBox  entry."
- "XStep and YStep may be either positive or negative but shall not be zero. "

yurydelendik replied:
... in case when Math.abs(XStep) is greater than cell width, we need to grow canvas, and in case when Math.abs(XStep) less than cell width we probably need to create a pattern that is painted on itself.

The canvas 2d context does not let us to set repeat distances so we probably need to do something clever, like paint in SVG, produce a canvas and "grow" it with SVG, disable "repeat" and repeat in fill/stroke. Also we need to apply optimizations for well-know paint patterns (e.g. painting simple image as using pattern) to replace with simple primitives.

yurydelendik commented on src/display/pattern_helper.js:
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
-      var tmpTranslate = [1, 0, 0, 1, -topLeft[0], -topLeft[1]];
-      graphics.transform.apply(graphics, tmpTranslate);
-
+      graphics.transform(dimx.scale, 0, 0, dimy.scale, 0, 0);
```
shall it have additional -x0, -y0 translate?

Rob--W replied:
No, I removed it because the transformation actually resulted in a bad rendering: fad83a7deb2d3cb1b6597f11e4e5ecfd7e9a3c94.

yurydelendik commented on src/display/pattern_helper.js:
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
-      width = Math.min(Math.ceil(Math.abs(width * combinedScale[0])),
-        MAX_PATTERN_SIZE);
+      var ctm = this.ctx.mozCurrentTransform;
+      var dimx = this.getSizeAndScale(xstep, this.ctx.canvas.width, ctm[0]);
```
Don't rely on ctm alone, use singularValueDecompose2dScale, that's why combinedScale was introduced.

yurydelendik commented on src/display/pattern_helper.js:
```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
-      this.ctx.scale(1 / scale[0], 1 / scale[1]);
+    getSizeAndScale:
+        function TilingPattern_getSizeAndScale(step, realOutputSize, scale) {
+      // TODO(robwu): The scale should not be 0. Is this the right place to
```
See comment above about singularValueDecompose2dScale, in your case the scale parameter will be zero when `ctx.rotate(Math.PI/2)` is used.

timvandermeij commented:
@Rob--W Do you have time to rebase this PR and fix the comments above? It looks like a nice patch to have as it fixes quite some open issues.

Rob--W commented:
@timvandermeij Last month I discussed this PR with @yurydelendik over IRC (http://logs.glob.uno/?c=mozilla%23pdfjs&s=13+Nov+2015&e=13+Nov+2015&h=6606#c41194) and the conclusion is that the implementation can be improved. This change is a significant refactor of the tiling and drawing code, and I haven't allocated time for that task yet. Of course, I can rebase and fix the nits that are mentioned above (so that we have the desired functionality) and submit a new PR for an improved version.

If @yurydelendik does not have any objections against merging this PR (with the nits fixed), then I will rebase and update the PR so you can merge it.

Snuffleupagus commented:
Given that this PR completely fixes a handful of issues, and also improves a few more, could we perhaps now reconsider landing this as is (with the last round of review comments addressed)!?
It seems to me that while it's certainly possible to improve the general implementation (as outlined in the IRC discussion), this PR would still be a nice improvement of the pattern handling.

timvandermeij commented:
I agree with the above. I do not know enough of this part of the codebase to review it, but if it's an improvement (perhaps not a perfect solution yet), I would also like to have this landed. It can always be improved later on.

janpe2 commented:
@Rob--W After addressing the review comments, the class TilingPattern could maybe look like this:
https://gist.github.com/janpe2/beb4a18d73dffb250830e7ddc41bd461
I have tested lots of PDFs and this would fix as many as 14 issues.

timvandermeij commented:
Let's get this PR rolling again since it will fix quite a large number of issues. @Rob--W Do you have time to fix this up based on the comment above? If not, perhaps @janpe2 could create a new PR with the suggested changes.

Snuffleupagus commented:
Given that there hasn't been any response to https://github.com/mozilla/pdf.js/pull/6606#issuecomment-455872532 for almost a month, and that the issues keep on coming, I cannot see the harm in @janpe2 (provided you want to, of course) finishing this PR so that it can land.

Rob--W commented:
I'm slowly working on this.

Neither my PR nor @janpe2's PR are suitable for merging, because of one glaring issue (which I will describe below). I locally revived the patch when https://github.com/mozilla/pdf.js/pull/6606#issuecomment-455872532 was posted, but was concerned about the unaddressed issues (in https://github.com/mozilla/pdf.js/pull/6606/files/fad83a7deb2d3cb1b6597f11e4e5ecfd7e9a3c94#r44318712).

To check whether there are any potential regressions, I was looking for a test case (PDF) that has a pattern where the xstep that is smaller than the bounding box's width (or equivalently, ystep and height), AND that repeats that pattern.
In the current version of the PR, the width/height of the pattern canvas is set to the xstep/ystep (whereas it was previously set to the pattern's bounding box). The expected common case is where `xstep > width` (or `ystep > height`). This is rendered well in the current PR.
In the case where `xstep < width` (or `ystep < height`) AND the pattern is repeated, part of the pattern may be cut off.

For example, a `V` pattern with the xstep set to half of its width.
```
Base pattern:
\/

Expected rendering of repeated pattern (recall xstep = half of width):
\XXXX

Actual rendering if truncating pattern width to xstep:
\\\\\
```

To check whether the issue is real, and to confirm a fix, I need an actual test case (PDF). I'm probably going to create such a test case by hand, since searching for an existing test case with this issue is like looking for a needle in a haystack.

THausherr commented:
@Rob--W 
[patterns-Xstep5.pdf](https://github.com/mozilla/pdf.js/files/2879161/patterns-Xstep5.pdf)
[patterns-Xstep10.pdf](https://github.com/mozilla/pdf.js/files/2879163/patterns-Xstep10.pdf)

THausherr commented:
@Rob--W 
[patterns-Xstep5.pdf](https://github.com/mozilla/pdf.js/files/2879161/patterns-Xstep5.pdf)
[patterns-Xstep10.pdf](https://github.com/mozilla/pdf.js/files/2879163/patterns-Xstep10.pdf)
[patterns-Xstep7.pdf](https://github.com/mozilla/pdf.js/files/2879197/patterns-Xstep7.pdf)

janpe2 commented:
This is what the standard says about overlapping tiles (section 8.7.3.1):
> The order in which individual tiles (instances of the cell) are painted is unspecified and unpredictable; figures on adjacent tiles should not overlap.

However, Acrobat Reader draws them with no problems.

---

If someone is interested, here are the issues that would be fixed by my proposed patch:
#6496
#1434
#2825
#8264
#7731
#9763
#9927
#9928
#9922
#8579
#9665
#8083
#8373
[Bugzilla-1467424](https://bugzilla.mozilla.org/show_bug.cgi?id=1467424)

Rob--W commented:
@THausherr  Thanks for sharing these test cases. That really saves me some time!
@janpe2  Thanks for checking those. I expect that the final version of this PR will fix those bugs, and avoid the regression that I described in https://github.com/mozilla/pdf.js/pull/6606#issuecomment-464784033

THausherr commented:
Here's another. The shape to the top and to the right should be "pointy" and not flat when rendering is correct.
[gs-bugzilla690487.pdf](https://github.com/mozilla/pdf.js/files/2881471/gs-bugzilla690487.pdf)

The file from #9770 does also have a "too small" step.

@Rob--W your comment from a day ago also showed that there's a bug in PDFBox 😭

Rob--W commented:
FYI I was and am still traveling, and will not update this PR before 16 March. Thanks for your patience.

Rob--W committed:
TilingPattern: Add comment to explain the implementation

Rob--W commented:
I have updated the PR, and while addressing the review I also tried to minimize the changes compared to the current master. I did not fix the issue that I described in https://github.com/mozilla/pdf.js/pull/6606#issuecomment-464784033 because the issue already existed before my PR (so it is not a regression), and addressing it is complicated and may result in worse performance.

Also, the specific test case that I described is not a valid PDF: Tiling patterns may overlap, but the figures on the pattern cells are not allowed to overlap by the PDF spec.

The test that I added passes, as well as the ones that I referenced in the original first comment on this PR. Please test whether this fixes other issues (@ contributors on this thread) and merge it if it looks good (@timvandermeij ).

timvandermeij commented:
Thank you for fixing this up! I see that Travis reports some linting issues; could you fix those? For line 412, you should be able to use the shorthand `return { scale, size, };` instead to solve those errors.

Rob--W commented:
@timvandermeij Done.

timvandermeij commented:
/botio test

timvandermeij commented:
I'm just wondering, should we add a test case for commit https://github.com/mozilla/pdf.js/pull/6606/commits/c4c92da57143ab39fe1c78760206a0e67a21e876 or is that also covered by the test case you have already added in another commit here?

Rob--W commented:
The test case of the third commit already covered the second patch mostly: The PDF consists of 63 patterns with bbox `[0 0 596 842]` (same as page size), and xstep = ystep = 2876. Before my second patch, the result would be blurry in Firefox and cause an out-of-memory crash in Chrome. With the patch, the rendering is sharp.

I added a new test anyway, a minimal file that tests the relevant edge cases that caused this bug. It also features another change compared to previous code: I'm using `Math.max(MAX_PATTERN_SIZE, realOutputSize)` instead of `Math.min`, because otherwise the pattern would be clipped if the output canvas size is larger than `MAX_PATTERN_SIZE` (e.g. when the user uses 400% zoom).

Rob--W commented:
/botio test

Rob--W commented:
I updated the first comment with a list of bugs that can be closed once this is merged, from the list of issues with the `4-tiling-type` or `4-shading-pattern` label, or were listed in an earlier comment:

- https://github.com/mozilla/pdf.js/labels/4-tiling-type
- https://github.com/mozilla/pdf.js/issues?q=is%3Aopen+label%3A4-shading-pattern+-label%3A4-tiling-type
- https://github.com/mozilla/pdf.js/pull/6606#issuecomment-465224776

timvandermeij commented:
/botio test

timvandermeij commented:
/botio-linux preview

timvandermeij reviewed (approved):
Nice work fixing this many issues! The commits are clear and the extra comments are helpful. The increased reference test coverage is also really good.

timvandermeij commented:
/botio makeref
